File: module/visual-jdbc/src/main/java/org/visual/jdbc/util/JDBCUtil.java
/* (C)2024*/
package org.visual.jdbc.util;

import java.sql.Connection;
import java.util.*;
import lombok.NonNull;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.visual.jdbc.core.DatabaseResultKey;
import org.visual.jdbc.core.DatabaseTableColumn;

@UtilityClass
public class JDBCUtil {

  @SneakyThrows
  public static @NotNull Set<String> listDatabase(@NonNull Connection conn) {
    val metadata = conn.getMetaData();
    val c = metadata.getCatalogs();
    val result = new HashSet<String>();
    while (c.next()) {
      result.add(c.getString(DatabaseResultKey.TABLE_CAT.getKey()));
    }
    return result;
  }

  @SneakyThrows
  public static @NotNull Set<String> listTables(@NonNull Connection conn) {
    val metadata = conn.getMetaData();
    val catalog = conn.getCatalog();
    val tables = metadata.getTables(catalog, null, null, new String[] {"TABLE"});
    val result = new HashSet<String>();
    while (tables.next()) {
      result.add(tables.getString(3));
    }
    return result;
  }

  @SneakyThrows
  public static @NotNull Set<DatabaseTableColumn> listColumns(
      @NonNull Connection conn, @NonNull String tableName) {
    val getColumns = conn.getMetaData().getColumns(conn.getCatalog(), null, tableName, null);
    val result = new HashSet<DatabaseTableColumn>();
    while (getColumns.next()) {
      val col =
          DatabaseTableColumn.builder()
              .columnName(getColumns.getString(DatabaseResultKey.COLUMN_NAME.getKey()))
              .typeName(getColumns.getString(DatabaseResultKey.TYPE_NAME.getKey()))
              .columnSize(getColumns.getInt(DatabaseResultKey.COLUMN_SIZE.getKey()))
              .nullable(getColumns.getInt(DatabaseResultKey.NULLABLE.getKey()))
              .build();
      result.add(col);
    }
    return result;
  }

  @SneakyThrows
  public static @NotNull Set<String> listPrimaryKeys(
      @NonNull Connection conn, @NonNull String tableName) {
    val getPrimaryKey = conn.getMetaData().getPrimaryKeys(conn.getCatalog(), null, tableName);
    val result = new HashSet<String>();
    while (getPrimaryKey.next()) {
      result.add(getPrimaryKey.getString(DatabaseResultKey.COLUMN_NAME.getKey()));
    }
    return result;
  }

  @SneakyThrows
  public Map<String, List<String>> listIndexes(@NotNull Connection conn, String tableName) {
    val metaData = conn.getMetaData();
    val getIndexInfo = metaData.getIndexInfo(null, null, tableName, false, true);

    Map<String, List<String>> result = new HashMap<>();

    while (getIndexInfo.next()) {
      String indexName = getIndexInfo.getString(DatabaseResultKey.INDEX_NAME.getKey());
      String columnName = getIndexInfo.getString(DatabaseResultKey.COLUMN_NAME.getKey());
      result.computeIfAbsent(indexName, k -> new ArrayList<>()).addFirst(columnName);
    }

    return result;
  }
}

File: visual-model-database/src/main/kotlin/org/visual/model/sql/JDBCUtil.kt
@file:JvmName("JDBCUtil")

package org.visual.model.sql

import java.sql.Connection

fun listDatabase(conn: Connection): Set<String> {
    val metadata = conn.metaData
    val c = metadata.catalogs
    return generateSequence {
        c
    }.takeWhile { c.next() }.map {
        c.getString(DatabaseResultKey.TABLE_CAT.key)
    }.toSet()
}

fun listTables(conn: Connection): Set<String> {
    val metadata = conn.metaData
    val catalog: String = conn.catalog
    val tables = metadata.getTables(catalog, null, null, arrayOf("TABLE"))
    return generateSequence {
        tables
    }.takeWhile { tables.next() }
        .map { tables.getString(3) }
        .toSet()
}

fun listColumns(conn: Connection, tableName: String): Set<DatabaseTableColumn> {
    val getColumns = conn.metaData.getColumns(conn.catalog, null, tableName, null)
    return generateSequence { getColumns }
        .takeWhile { it.next() }
        .map {
            DatabaseTableColumn.builder()
                .columnName(it.getString(DatabaseResultKey.COLUMN_NAME.key))
                .typeName(it.getString(DatabaseResultKey.TYPE_NAME.key))
                .columnSize(it.getInt(DatabaseResultKey.COLUMN_SIZE.key))
                .nullable(it.getInt(DatabaseResultKey.NULLABLE.key))
                .build()
        }
        .toSet()
}

fun listPrimaryKeys(conn: Connection, tableName: String): Set<String> {
    val metaData = conn.metaData
    val getPrimaryKey = metaData.getPrimaryKeys(conn.catalog, null, tableName)
    return generateSequence { getPrimaryKey }
        .takeWhile { it.next() }
        .map { it.getString(DatabaseResultKey.COLUMN_NAME.key) }
        .toSet()
}

fun listIndexes(conn: Connection, tableName: String): Map<String, List<String>> {
    val metaData = conn.metaData
    val getIndexInfo = metaData.getIndexInfo(null, null, tableName, false, true)
    return generateSequence { getIndexInfo }
        .takeWhile { it.next() }
        .fold(mutableMapOf()) { acc, resultSet ->
            val indexName = resultSet.getString(DatabaseResultKey.INDEX_NAME.key)
            val columnName = resultSet.getString(DatabaseResultKey.COLUMN_NAME.key)
            acc.computeIfAbsent(indexName) { mutableListOf() }.addFirst(columnName)
            acc
        }
}